### PR TITLE
fixed disabled guest checkout issue in case of downloadable product 

### LINKED ISSUE #23971: Disable Guest Checkout if Cart Contains Downloadable Items doesn't work correctly
### Preconditions (*)

1. Magento version 2.3.2
2. Php version: 7.2.19

### Steps to reproduce (*)

1. navigate to backend->stores->configuration->catalog->Downloadable Product Options
2. Set Shareable -> No and Disable Guest Checkout if Cart Contains Downloadable Items -> No
downloadable-product-issue

### Expected result (*)

1. Guest checkout must be disabled if cart contains downloadable products.

### Actual result (*)

1. Guest checkout is enabled

### PR DESCRIPTION
**Pull request by rani-priya**
fixed disabled guest checkout issue in case of downloadable product #23971

### Related Pull Requests
* https://github.com/magento/inventory/pull/2966

### Fixed Issues (if relevant)

1. fixes magento/magento2#23971: Disable Guest Checkout if Cart Contains Downloadable Items doesn't work correctly

### Manual testing scenarios (*)

1. navigate to backend->stores->configuration->catalog->Downloadable Product Options
2. Set Shareable -> No and Disable Guest Checkout if Cart Contains Downloadable Items -> No
downloadable-product-issue

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by josefbehr** [comment]
@rani-webkul apparently, some tests have to be adapted, too, for this to work. Please fix those, and I see no problem in merging this. Thank you!

**Comment by josefbehr** [comment]
@rani-webkul Awesome, thanks.

**Review by josefbehr — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @josefbehr, thank you for the review.
[ENGCOM-5629](https://jira.corp.magento.com/browse/ENGCOM-5629) has been created to process this Pull Request
:eight_spoked_asterisk: @josefbehr, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

File: app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php [hunk 1]
```diff
@@ -8,13 +8,21 @@
...
 {
     /**
      *  Xml path to disable checkout
      */
     const XML_PATH_DISABLE_GUEST_CHECKOUT = 'catalog/downloadable/disable_guest_checkout';
 
+    /**
+     *  Xml path to get downloadable Shareable setting
+     */
+    const XML_PATH_DOWNLOADABLE_SHAREABLE = 'catalog/downloadable/shareable';
```
Comment (sidolov) [inline app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php]: Please, make const private

**Review by sidolov — CHANGES_REQUESTED** [review]

File: app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php (hunk 1 above)
Comment (rani-priya) [reply app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php]: done

**Review by josefbehr — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @josefbehr, thank you for the review.
[ENGCOM-5629](https://jira.corp.magento.com/browse/ENGCOM-5629) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
Hi @rani-webkul 

During testing we faced with the issue

**_Problem:_** We are not able to proceed checkout as guest when we set value "Shareable"="Yes" during creating downloadable product

**Steps to reproduce:**

1. Login to Admin;
2. Navigate to Stores->Configuration->Catalog->Catalog->Downloadable Product Options;
3. Set:
- Shareable = No;
- Disable Guest Checkout if Cart Contains Downloadable Items = No;
4. Create Downloadable product and in "Download Information" tab set "Shareable" to "Yes"; 
screen3
5. Go to Storefront as guest;
6. Add created before product to Cart;
7. Click on "Proceed to Checkout";

**Actual Result:**
Guest checkout is disabled
screen4

**Expected Result:**
Guest checkout must be enabled if cart contains downloadable products.

@rani-webkul Could you take a look, please?

Thanks!

**Review by sidolov — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sidolov, thank you for the review.
[ENGCOM-5629](https://jira.corp.magento.com/browse/ENGCOM-5629) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
Hi @rani-webkul, the issue described in [#comment](https://github.com/magento/magento2/pull/23972#issuecomment-527845043) is still reproduce.

@rani-webkul Could you take a look, please? Thanks!

File: app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php (hunk 1 above)
Comment (sidolov) [reply app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php]: Please, make const private again

File: app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php [hunk 2]
```diff
@@ -15,20 +15,34 @@ class IsAllowedGuestCheckoutObserver implements ObserverInterface
...
      * @var \Magento\Framework\App\Config\ScopeConfigInterface
      */
     protected $_scopeConfig;
 
+    /**
+     * Downloadable link collection factory
+     *
+     * @var \Magento\Downloadable\Model\ResourceModel\Link\CollectionFactory
+     */
+    protected $_linksFactory;
```
Comment (sidolov) [inline app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php]: Please, make property private and avoid underscore in property name

File: app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php [hunk 3]
```diff
@@ -15,20 +15,34 @@ class IsAllowedGuestCheckoutObserver implements ObserverInterface
...
+
     /**
      * @param \Magento\Framework\App\Config\ScopeConfigInterface $scopeConfig
      */
     public function __construct(
-        \Magento\Framework\App\Config\ScopeConfigInterface $scopeConfig
+        \Magento\Framework\App\Config\ScopeConfigInterface $scopeConfig,
+        \Magento\Downloadable\Model\ResourceModel\Link\CollectionFactory $linksFactory
```
Comment (sidolov) [inline app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php]: Please, inject new dependencies according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

**Review by sidolov — CHANGES_REQUESTED** [review]

**Comment by engcom-Foxtrot** [comment]
@sidolov @engcom-Alfa updated.

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by ihor-sviziev** [comment]
Hi @engcom-Foxtrot,
There still test failures, could you take a look?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please check test failures

**Comment by ihor-sviziev** [comment]
Hi @engcom-Foxtrot @engcom-Charlie,
Could you check why tests started failing again?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @engcom-Foxtrot @engcom-Charlie,
Could you check why tests started failing again?

**Comment by engcom-Kilo** [comment]
Failed functional tests not related to the changes in this PR

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @rani-webkul or @engcom-Kilo,
Could you fix failing MFTF test?
You can find failure here https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/23972/632d9565286cc6e53b447e189ffd3b63/Functional/allure-report-ce/index.html#categories/c359ef4046b3e42effd5a524b6a9806b/a282c4cb90633b1e/

**Comment by engcom-Kilo** [comment]
Added Pull Request for MSI magento/inventory#2961 with fix for MFTF test.
As Pull Request will approved and merged, MFTF tests will be passed.

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php [hunk 4]
```diff
@@ -3,65 +3,121 @@
...
     /**
      *  Xml path to disable checkout
      */
-    const XML_PATH_DISABLE_GUEST_CHECKOUT = 'catalog/downloadable/disable_guest_checkout';
+    private const XML_PATH_DISABLE_GUEST_CHECKOUT = 'catalog/downloadable/disable_guest_checkout';
+
+    /**
+     *  Xml path to get downloadable Shareable setting
+     */
+    private const XML_PATH_DOWNLOADABLE_SHAREABLE = 'catalog/downloadable/shareable';
```
Comment (lbajsarowicz) [inline app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php]: `const` name is descriptive-enough, you don't need to add docblock
```suggestion
    private const XML_PATH_DOWNLOADABLE_SHAREABLE = 'catalog/downloadable/shareable';
```

File: app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php [hunk 5]
```diff
@@ -3,65 +3,121 @@
...
+/**
+ * Checks if guest checkout is allowed then quote contains downloadable products.
+ */
 class IsAllowedGuestCheckoutObserver implements ObserverInterface
 {
     /**
      *  Xml path to disable checkout
      */
-    const XML_PATH_DISABLE_GUEST_CHECKOUT = 'catalog/downloadable/disable_guest_checkout';
+    private const XML_PATH_DISABLE_GUEST_CHECKOUT = 'catalog/downloadable/disable_guest_checkout';
```
Comment (lbajsarowicz) [inline app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php:25]: You can also remove docblock here, as it does not introduce any value.

File: app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php [hunk 6]
```diff
@@ -3,65 +3,121 @@
...
-     * @var \Magento\Framework\App\Config\ScopeConfigInterface
+     * @var ScopeConfigInterface
      */
-    protected $_scopeConfig;
+    private $scopeConfig;
 
     /**
-     * @param \Magento\Framework\App\Config\ScopeConfigInterface $scopeConfig
+     * Downloadable link collection factory
+     *
```
Comment (lbajsarowicz) [inline app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php]: We don't need extra description here. `@var` is enough

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Could you ammend your PR following my comments, please? :-)

File: app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php (hunk 5 above)
Comment (ihor-sviziev) [reply app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php:25]: Removed

File: app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php (hunk 6 above)
Comment (ihor-sviziev) [reply app/code/Magento/Downloadable/Observer/IsAllowedGuestCheckoutObserver.php]: Removed

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by lbajsarowicz — APPROVED** [review]
:heavy_check_mark: Thank you for your contribution

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-5629](https://jira.corp.magento.com/browse/ENGCOM-5629) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@lbajsarowicz I'll keep it in "in review" till tests will be passing, also just added related mftf fix in inventory repo.

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by ihor-sviziev** [comment]
@magento run Integration Tests, Unit Tests

**Comment by ihor-sviziev** [comment]
@magento run all test

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — DISMISSED** [review]
Approved ✔
Failing unit, integration, functional and db compare tests are not related to changes from this PR

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-5629](https://jira.corp.magento.com/browse/ENGCOM-5629) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

### _What was tested:_

1. _**The problem described in #23971**_ 

**Manual testing scenario:**

1. Create Downloadable product with Shareable set no **No**
Screenshot from 2020-04-17 16-02-42
2. Go to Admin -> Stores -> Configuration -> Catalog -> Catalog -> Downloadable Product Options;
3.  Set `Shareable -> No` & ` Disable Guest Checkout if Cart Contains Downloadable Items -> No`
screenshot4444444

**Before:** :heavy_multiplication_x: Guest checkout is enabled

screenshotbefore

**After:** :heavy_check_mark:  Guest checkout is disabled 

Screenshot from 2020-04-17 16-38-27


_**2. Also checked the additional cases:**_

- Downloadable product with Shareable = **Yes**

**Manual testing scenario:**

1. Create Downloadable product with Shareable set to **Yes**;
2. Go to Admin -> Stores -> Configuration -> Catalog -> Catalog -> Downloadable Product Options;
3.  Set `Shareable -> No` & ` Disable Guest Checkout if Cart Contains Downloadable Items -> No`
screenshot4444444

**Actual Result:** :heavy_check_mark: Guest checkout is enabled

screenshotbefore

- Downloadable product with Shareable = **Use config**

**Manual testing scenario:**

1. Create Downloadable product with Shareable set to **Use config**;
2. Go to Admin -> Stores -> Configuration -> Catalog -> Catalog -> Downloadable Product Options;
3.  Set `Shareable -> No` & ` Disable Guest Checkout if Cart Contains Downloadable Items -> No`
screenshot4444444

**Actual Result:** :heavy_check_mark: Guest checkout is disabled

Screenshot from 2020-04-17 16-38-27

- **Disable Guest Checkout if Cart Contains Downloadable Items** set to **Yes**

**Manual testing scenario:**

1. Create Downloadable product with Shareable set to **Use config**;
2. Go to Admin -> Stores -> Configuration -> Catalog -> Catalog -> Downloadable Product Options;
3.  Set `Shareable -> Yes` & ` Disable Guest Checkout if Cart Contains Downloadable Items -> Yes`
screenshot777

**Actual Result:** :heavy_check_mark: Guest checkout is disabled

Screenshot from 2020-04-17 16-38-27

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by engcom-Kilo** [comment]
Failed Database Compare test not related to the changes in this PR

**Review by slavvka — CHANGES_REQUESTED** [review]
@rani-webkul please resolve conflicts and rerun tests after upgreade to PHP 7.4

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @rani-webkul,
I fixed conflicts myself, but now unit test started failing. Could you fix it?

**Comment by ihor-sviziev** [comment]
Hi @rani-webkul,
Will you be able to update your PR?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Unit Tests

**Comment by ihor-sviziev** [comment]
@magento run Unit Tests

**Comment by ihor-sviziev** [comment]
@magento run Unit Tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]
✔ Approved.

Failing tests looks not related to changes form this PR.

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

### _What was tested:_

1. _**The problem described in #23971**_ 

**Manual testing scenario:**

1. Create Downloadable product with Shareable set no **No**
Screenshot from 2020-04-17 16-02-42
2. Go to Admin -> Stores -> Configuration -> Catalog -> Catalog -> Downloadable Product Options;
3.  Set `Shareable -> No` & ` Disable Guest Checkout if Cart Contains Downloadable Items -> No`
screenshot4444444

**Before:** :heavy_multiplication_x: Guest checkout is enabled

screenshotbefore

**After:** :heavy_check_mark:  Guest checkout is disabled 

Screenshot from 2020-04-17 16-38-27


_**2. Also checked the additional cases:**_

- Downloadable product with Shareable = **Yes**

**Manual testing scenario:**

1. Create Downloadable product with Shareable set to **Yes**;
2. Go to Admin -> Stores -> Configuration -> Catalog -> Catalog -> Downloadable Product Options;
3.  Set `Shareable -> No` & ` Disable Guest Checkout if Cart Contains Downloadable Items -> No`
screenshot4444444

**Actual Result:** :heavy_check_mark: Guest checkout is enabled

screenshotbefore

- Downloadable product with Shareable = **Use config**

**Manual testing scenario:**

1. Create Downloadable product with Shareable set to **Use config**;
2. Go to Admin -> Stores -> Configuration -> Catalog -> Catalog -> Downloadable Product Options;
3.  Set `Shareable -> No` & ` Disable Guest Checkout if Cart Contains Downloadable Items -> No`
screenshot4444444

**Actual Result:** :heavy_check_mark: Guest checkout is disabled

Screenshot from 2020-04-17 16-38-27

- **Disable Guest Checkout if Cart Contains Downloadable Items** set to **Yes**

**Manual testing scenario:**

1. Create Downloadable product with Shareable set to **Use config**;
2. Go to Admin -> Stores -> Configuration -> Catalog -> Catalog -> Downloadable Product Options;
3.  Set `Shareable -> Yes` & ` Disable Guest Checkout if Cart Contains Downloadable Items -> Yes`
screenshot777

**Actual Result:** :heavy_check_mark: Guest checkout is disabled

Screenshot from 2020-04-17 16-38-27

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Pull request merged**